lib/Dialect/FIRRTL/Transforms/InnerSymbolDCE.cpp

```diff
@@ -0,0 +1,106 @@
+//===- InnerSymbolDCE.cpp - Delete Unused Inner Symbols----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass removes inner symbols which have no uses.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct InnerRefInfo {
+  unsigned refcount = 0;
+  Operation *op;
+};
+} // namespace
+
+struct InnerSymbolDCEPass : public InnerSymbolDCEBase<InnerSymbolDCEPass> {
+  void runOnOperation() override;
+};
+
+void InnerSymbolDCEPass::runOnOperation() {
+  DenseMap<hw::InnerRefAttr, InnerRefInfo> innerRefMap;
+
+  CircuitOp circuit = getOperation();
+
+  auto examineAnnotation = [&](Annotation anno) -> void {
+    anno.getDict().walkSubAttrs([&](Attribute attr) {
+      auto innerRef = attr.dyn_cast<hw::InnerRefAttr>();
+      if (!innerRef)
+        return;
+      innerRefMap[innerRef].refcount++;
+    });
+  };
+
+  auto examineAnnotations = [&](Operation *op) -> void {
+    for (auto anno : AnnotationSet(op))
+      examineAnnotation(anno);
+  };
+
+  examineAnnotations(circuit);
+  for (Operation &op : circuit.getBody()->getOperations()) {
+    examineAnnotations(&op);
+
+    // Extract any hw::InnerRefAttrs from NLAs.
+    if (auto nla = dyn_cast<NonLocalAnchor>(op)) {
+      for (auto path : nla.namepath()) {
+        auto innerref = path.dyn_cast<hw::InnerRefAttr>();
+        if (!innerref || !innerRefMap.count(innerref))
+          continue;
+        innerRefMap[innerref].refcount++;
```

Comment:
The `!innerRefMap.count(innerref)` feels like it could cause the pass to remove symbols even if there is an NLA referencing it. What happens if you have something like:
```mlir
firrtl.nla @foo [#hw.innerNameRef<@Foo::@bar>]
firrtl.module @Foo() {
  firrtl.wire sym @bar {inner_sym_visibility = "private"} : !firrtl.uint<1>
}
```
When that NLA is visited, wouldn't the `innerRefMap` count be zero, preventing the NLA from incrementing the refcount, which would later delete the symbol even though it's used by the NLA?

Reply:
This is an old comment, but the current version should respect NLAs!